[Lim Bing Sen] iP

bingsen0806 opened the pull request:
## Charizard #123 
> "Gotta Burn Em All~" _by Charizard_ ([Source](https://www.youtube.com/watch?v=2WVAIlueaH8)) 

* Super ~~_fast_~~ ***hot***
* Text-based
* ~~Keeps track~~ ***Burns away*** your tasks!

Steps for summoning Charizard:
1. Download the `jar` file at this [link](https://github.com/bingsen0806/ip/releases/download/A-Jar/main.jar).
2. Double click it.
3. Use `list` command to view your _burning list_.
4. Use `add` command to add tasks.
5. _And finally ..._ `mark` to **burn away tasks**!

### Features
- [X] Managing tasks 📖  
- [X] Managing deadlines 📅 
- [ ] Chat-style GUI (coming soon) 💻  

### For developers
If you are a Java developer, feel free to take a look at the code. The `main` method is as follows:

```
public Class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
### Happy Burning
![Charizard](https://static.pokemonpets.com/images/monsters-images-300-300/6-Charizard.webp)

Commit:
Add Gradle support

bingsen0806 committed:
Add functionality to read and echo output.

bingsen0806 committed:
Move all code from Main.java to Duke.java

bingsen0806 committed:
Add functionality to add task and list task.

bingsen0806 committed:
Add functionality to mark task as Done or Undone.

bingsen0806 committed:
Add 3 new types of task: ToDo, Deadline and Event.
Change Task to abstract class.

bingsen0806 committed:
Add input and expected output files for automated text UI testing

bingsen0806 committed:
Add exceptions to handle invalid inputs for wrong command, invalid todo command, mark and unmark command, deadline command, event command.
Update expected output for automated text-ui testing.
Abstract out error handling logic into seprate methods for each command.

bingsen0806 committed:
Move all replies from Duke as final static String variables.

bingsen0806 committed:
Add delete functionality

bingsen0806 committed:
Modify Task class to use TaskType enum

bingsen0806 committed:
Duke file: Save tasks to harddrive

Tasks only exists throughout the lifetime of programme.

Non-persistent tasks does not allow users to retrieve data stored during
previous execution of the programme.

Let's,
* initialize the task list from a file on program load
* save the task list to a file on exit
* use relative path to track the location of the save file

Using relative path will allow our programme to be platform independent
and be able to run on different machines regardless of where the program
is downloaded.

bingsen0806 committed:
Task: Modidy Deadline and Event to use LocalDateTime instead of String

Deadline and event uses String to store their deadline and event time.

Such strings are not validated and may not contain information regarding
date and time

Let's
* Use LocalDateTime instead of String to represent event and deadlines
* Validate the user String input to check for correct datetime format

Validating datetime input ensures that Deadline and Event objects stores
the correct time for the respective tasks.

bingsen0806 committed:
Merge branch 'branch-Level-7'

bingsen0806 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

bingsen0806 committed:
Refactor code using OOP principles

bingsen0806 committed:
Refactor code in packages and changed method accessibility.

bingsen0806 committed:
ParserTest and TaskListTest: Add JUnit test

bingsen0806 committed:
Add jar file and modify gitignore to exclude META-INF directory

bingsen0806 committed:
Add Javadocs for Parser class methods.

bingsen0806 committed:
Add JavaDocs

bingsen0806 committed:
Edit code to adhere to coding standards.

bingsen0806 committed:
Add Find functionality to find tasks based on substring of description

bingsen0806 committed:
Merge branch 'branch-A-JavaDoc'

bingsen0806 committed:
Merge branch 'branch-A-CodingStandard'

bingsen0806 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/main/Parser.java

bingsen0806 committed:
Update JavaDocs related to find functionality after merging

bingsen0806 committed:
Merge branch 'add-gradle-support' of https://github.com/bingsen0806/ip into add-gradle-support

bingsen0806 committed:
Merge branch 'add-gradle-support'

Ashuh commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,19 @@
+    public static final String ERROR_TODO_NO_NAME = "Please specify the name of new task to be burnt";
+    public static final String ERROR_FIND_NO_NAME = "Please specify the name of task to find";
+    public static final String ERROR_WRONG_FORMAT = "Charizard remembers this move but cannot recognize the format.";
+    public static final String ERROR_IO_Input = "Unable to read from input.";
```
Should the variable name be fully capitalized?

bingsen0806 committed:
Add gradle support to project

bingsen0806 committed:
Merge branch 'branch-A-Gradle'

Ashuh commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,48 @@
+        isDone = false;
+    }
+
+    public void mark(boolean done) {
+        isDone = done;
+    }
```
Should boolean argument names be prefixed with is/has/was?

Ashuh replied:
Should the name of this method be `setDone` instead?

bingsen0806 committed:
Modify files to follow checkstyles.

bingsen0806 committed:
Merge branch 'branch-A-CheckStyle'

Ashuh commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,105 @@
+     * If the file does not exist, returns an empty ArrayList.
+     * @throws DukeException If file input cannot be read of file input has wrong format.
+     */
+    public ArrayList<Task> loadTasks() throws DukeException {
+        File file = new File(filePath);
+        if (!file.exists()) {
+            return new ArrayList<Task>();
+        }
+        try {
+            BufferedReader fileReader = new BufferedReader(new FileReader(filePath));
+            String taskInput;
+            ArrayList<Task> readTasks = new ArrayList<Task>();
+            while ((taskInput = fileReader.readLine()) != null) {
+                String[] splitTaskInput = taskInput.split(" \\| ");
+                Task newTask;
+                if (splitTaskInput[0].equals("T")) {
+                    newTask = new ToDo(splitTaskInput[2]);
+                } else if (splitTaskInput[0].equals("D")) {
+                    LocalDateTime dateTime = Parser.parseDateTime(splitTaskInput[3]);
+                    newTask = new Deadline(splitTaskInput[2], dateTime);
+                } else if (splitTaskInput[0].equals("E")) {
+                    LocalDateTime dateTime = Parser.parseDateTime(splitTaskInput[3]);
+                    newTask = new Event(splitTaskInput[2], dateTime);
+                } else {
+                    throw new DukeException("Error: duke.task.Task type is not T,D or E in file\n");
+                }
+                int i = Integer.parseInt(splitTaskInput[1]);
+                if (i == 1) {
+                    newTask.mark(true);
+                }
+                readTasks.add(newTask);
+            }
+            return readTasks;
+        } catch (IOException e) {
+            throw new DukeException("Error: File input cannot be read\n" + e.getMessage());
+        } catch (NumberFormatException e) {
+            throw new DukeException("Error: isDone field is not indicated by 0 or 1 in file\n" + e.getMessage());
+        }
+    }
```
You might want to add some blank lines to separate logical units.

Ashuh commented on src/main/java/duke/main/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+     * @param searchDescription Search text to be partially matched with task's description.
+     * @return ArrayList of tasks that matches the description.
+     */
+    public ArrayList<Task> findTasks(String searchDescription) {
+        return new ArrayList<Task>(tasks.stream()
+                .filter(task -> task.hasSubstring(searchDescription))
+                .collect(Collectors.toList()));
+    }
```
I like how you used streams here!

Ashuh reviewed (commented):
Looks good! I could only find a few minor issues.

bingsen0806 committed:
Add some initial gradle configuration and basix fxml files for JavaFX.

bingsen0806 committed:
Lauch UI successfully without logical output.

bingsen0806 committed:
Improve GUI

bingsen0806 committed:
Merge branch 'branch-Level-10'

bingsen0806 committed:
Repackage into ui and utils package

bingsen0806 committed:
Build jar files for production.

bingsen0806 committed:
Package command and ui: use assertions

Assertions are not used to check whether the ui buffer is empty before a
new command is executed. Assertions are also not used to check whether
the program has exited.

Using assertions allows us to explicitly verify that the ui buffer is
empty at the state of program execution before new messages for a new
command is added. It also helps us ensure no response is received after
the program has exited.

Let's,
* Include a new function in Ui to check for empty buffer
* Use assertions and the new function to ensure the buffer is empty
before execution of any command
* Use assertions to ensure program does not receive response after it
has exited

bingsen0806 committed:
Refactor to improve code quality

There are several methods in Parser, Storage and MainWindow class that
are too long. There are also multiple magic string literals representing
error messages across different files. This makes it hard to change
locate usages of exisiting error messages and change them. The same
problem applies to starting and ending messages resulting from ui
commands.

Extracting all error messages into DukeException static variable allows
us to easily locate all error messages when they need to be changed.
Extracting parts of longer methods into shorter private methods also
improves readability of code and helps conform to SLAP principle.

Let's
* Extract methods that are longer than 30 LoC into smaller methods
* Extract all error messages into DukeException class as static class
variables
* Extract all magic string literals in command package to become static
variables of their respective classes

bingsen0806 committed:
Merge branch 'branch-A-Assertions'

bingsen0806 committed:
Merge branch 'branch-A-CodeQuality'

bingsen0806 committed:
Add stream tag for streams used in previous TaskList class

bingsen0806 committed:
Integrate CI into project

Project does not have CI integrated. There is a need to manually run
tests and during push.

Using GitHub Actions for CI allows us to specify jobs that needs to be
done in an automated manner when we push. This reduces time needed to
manually integrate code.

Let's
* Configure GitHub Actions to check for our code when we push

bingsen0806 committed:
Delete workflow

bingsen0806 committed:
Re-add workflow to test github authentication

bingsen0806 committed:
Delete text-ui test from yml file

bingsen0806 committed:
Update JUnit tests.

ian-from-dover commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,24 @@
+    public static final String ERROR_WRITE_FILE = "Unable to write to file";
+    public static final String FORMAT_DEADLINE = "Try using:\n\"deadline <task_name> /by <deadline>\".";
+    public static final String FORMAT_EVENT = "Try using:\n\"event <task_name> /by <event_time>\".";
+    public static final String FORMAT_DATE = "Date and time must be in yyyy/MM/dd HHmm format.";
```
I was expecting the error messages to be used within the class, or subclasses to be created for each of them. Is there a benefit to declaring them as public constants?

ianfromdover commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,106 @@
+
+    private static final int BORDER_LENGTH = 38;
+    private static final String GREET_MESSAGE = "Roarrr.... I'm Burning Charizard, tasked to burnnn down your tasks.\n"
+            + "Which task shall we burn today?";
```
Good placement of `+` operator 👍

ianfromdover commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,106 @@
+    public static final String REMOVE_MESSAGE = "Charizard got tired of waiting and deleted this task";
+    public static final String MARK_MESSAGE = "Charizard breathe out fire and burned the task.";
+    public static final String UNMARK_MESSAGE = "Oh no! The task was not burnt completely!";
+    public static final Image GENERAL_IMAGE = new Image(Ui.class.getResourceAsStream("/images/General.png"));
```
Neat and intuitive organisation and naming 👍

ianfromdover commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,106 @@
+        this.message.append(message);
+    }
+
+    /**
+     * Appends the line border to the stored message in Ui.
+     */
+    public void appendBorder() {
+        appendMessage("\n");
+        for (int i = 0; i < BORDER_LENGTH; i++) {
+            appendMessage("-");
+        }
+        appendMessage("\n");
+    }
```
Perhaps you might want to consider using `System.lineSeparator()` because on Windows a new line is `\r\n` while on Unix it is `\n`. A helpful way to implement it is to declare it as a constant such as `public static final String LS = System.lineSeparator();` along with your messages above.

bingsen0806 committed:
Add functionality for priority and allow sorting based on priority

bingsen0806 committed:
Merge branch 'branch-C-Priority'

ianfromdover commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,139 @@
+        } else if (fullCommand.startsWith("find")) {
+            String inputWithoutCommand = fullCommand.replaceFirst("find", "").trim();
+            return parseFindCommand(inputWithoutCommand);
+        } else {
+            throw new DukeException(DukeException.ERROR_NO_COMMAND);
+        }
```
Very clear way of organising error messages throughout the whole file 👍

bingsen0806 committed:
Add JavaDocs to SortCommand class

ianfromdover commented on text-ui-test/EXPECTED..txt:
```diff
@@ -0,0 +1,170 @@
+Hello from
+ ____        _        
+|  _ \ _   _| | _____ 
+| | | | | | | |/ / _ \
+| |_| | |_| |   <  __/
+|____/ \__,_|_|\_\___|
```
Perhaps you might want to generate custom ASCII art from a [website like this one](https://patorjk.com/software/taag/#p=display&h=0&v=0&f=Delta%20Corps%20Priest%201&t=Charizard).

ianfromdover reviewed (commented):
Overall well organised and I found no errors or style violations. Enjoyable to read.

bingsen0806 committed:
Merge pull request #2 from bingsen0806/branch-C-Priority

Add JavaDocs to SortCommand class

bingsen0806 committed:
Merge branch 'master' of https://github.com/bingsen0806/ip

bingsen0806 committed:
Add ui screenshot

bingsen0806 committed:
README.md : Add User Guide

bingsen0806 committed:
Add expected error messages to User Guide

bingsen0806 committed:
Add table of contents and image to User Guide

bingsen0806 committed:
Rename Ui.png.PNG to Ui.png

bingsen0806 committed:
Change border length to prevent border overflow for Mac users.
